coco/state.py

```diff
@@ -165,18 +178,56 @@ def read_from_file(self, path, file):
         file : str
             Name of the file to read from.
         """
-        logger.debug(f"Loading state {path} from file {file}.")
+        logger.debug(f"Loading file {file} into state path '{path}'.")
 
         # Update persistent state
         with self._storage.update():
-            element, name = self._find_new(path)
+            if len(path) != 0:
+                element, name = self._find_new(path)
 
             with open(file, "r") as stream:
                 try:
-                    element[name] = yaml.safe_load(stream)
+                    new_state = yaml.safe_load(stream)
+
+                    # Don't load state parts that are excluded from reset
+                    self._exclude_paths(path, new_state)
+
+                    if len(path) == 0:
+                        self._storage.state = new_state
+                    else:
+                        element[name] = new_state
                 except yaml.YAMLError as exc:
                     logger.error(f"Failure reading YAML file {file}: {exc}")
 
+    def _exclude_paths(self, path, state):
+        """
+        Remove excluded paths from a state (in-place).
+
+        If the excluded paths are set to `foo/bar` and `path` is `foo`, a this function would take
+        a `state = {'bar': 0}` and make it a `state = {}`.
+
+        Parameters
+        ----------
+        path : str
+            Prefix to 'state' when looking for parts to exclude.
+        state : dict
+            The state to remove excluded parts from.
+        """
+        if not isinstance(state, dict):
+            return
+        for excluded in self.exclude_from_reset:
+            if excluded.startswith(path):
+                if len(path) == 0:
+                    excluded = excluded[len(path) :]
+                else:
+                    # remove the `/` as well
+                    excluded = excluded[len(path) + 1 :]
+                path_first = excluded.split("/")[0]
+                if path_first in state:
+                    self._exclude_paths(path_first, state[path_first])
+                if excluded in state:
+                    del state[excluded]
```

Comment:
You can do all this in-place because python allows you to pass references to dictionaries?

Reply:
yes, right?

Reply:
Yeah, that is right, it just always catches me off-guard!

Are `dict`s the only object this applies to?

Reply:
Ja, it is confusing https://docs.python.org/3/faq/programming.html#how-do-i-write-a-function-with-output-parameters-call-by-reference